server/src/main/java/org/opensearch/querygroup/LabelingService.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.querygroup;
+
+import org.opensearch.action.IndicesRequest;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.plugins.LabelingPlugin;
+
+import java.util.EnumMap;
+import java.util.List;
+
+/**
+ * TODO: Don't know the right package to put this class, need suggestions for maintainers on this
+ * Main class to hold multiple implementations of {@link org.opensearch.plugins.LabelingPlugin}
+ * this class will facilitate access and interactions to different implementations
+ * Usage: This class should be used as a member to orchestrate the working of your {@link LabelingPlugin}
+ */
+public class LabelingService {
+    /**
+     * Enum to define what all are currently implementing the plugin
+     */
+    public enum LabelingImplementationType {
+        QUERY_GROUP_RESOURCE_MANAGEMENT,
+        NOOP
+    }
+
+    EnumMap<LabelingImplementationType, LabelingPlugin> implementations;
+
+    public LabelingService(List<LabelingPlugin> loadedPlugins) {
+        implementations = new EnumMap<>(LabelingImplementationType.class);
+        for (LabelingPlugin plugin : loadedPlugins) {
+            if (implementations.containsKey(plugin.getImplementationName())) {
+                throw new IllegalArgumentException("There should not be two implementations of a LabelingImplementation type");
+            }
+            implementations.put(plugin.getImplementationName(), plugin);
+        }
+    }
+
+    /**
+     * populates the threadContext with the labels yielded by the {@param type} against the {@link LabelingHeader} keys
```

Comment:
We are delegating the logic to "add labels into thread context" to plugin implementation, which IMO is a little bit dangerous. Plugins can add their arbitrary labels into thread context without complying with the `LabelingHeader` enum. I think we should have the logic to check/validate and put labels into thread context in this service, plugins should only compute the labels they want to add. Like: https://github.com/opensearch-project/OpenSearch/pull/14282/files#diff-f02ca59c8d99a70fdc277b1b311ba8fc0135920ef5a114578187e9388e922f6bR66

Reply:
I agree with this, I think with return types we can easily enforce the header types as key.